### ISSUE TITLE
Rethinking the userguide?

### ISSUE THREAD

**Issue by uliska** [frescobaldi/frescobaldi#1170]
I'm about to write documentation for the Document Fonts dialog and therefore got in touch with the `userguide` module again. Which makes me think whether we should reconsider it once more.

The main issue I have is that it doesn't seem very scalable in its current form. The least I would ask for is having the page hierarchy represented not manually by `#SUBDOCS` but by a physical directory structure. In general I'd say we have way too little documentation, so ideally there would be a massive amount of additional files, and it's somewhat confusing to look for stuff already now.

The other issue I'm wondering about is the actual templating/rendering mechanism used. Basically what we have is a selfwritten templating engine/site generator, and I think that's something one might rather want to avoid generally. It means we have to maintain everything ourselves, the functionality is limited to what we decide to implement ourselves, and it's not like it would look totally awesome in exchange ...

Wouldn't it be better to make use of established tools for the purpose? I don't think it is really necessary to always render the userguide pages on-the-fly since the manual is actually a static website displayed in a dialog. So I think it would work well to use an established site generator/documentation builder. The underlying language would be a secondary issue since it would only be necessary to be run on developers' computers, although it might make much sense to use a Python-based tool.

This would also make it possible to separate design and content a little easier and e.g. use nice HTML templates for the page layout and navigation structure, as well as alternative templates for 

* the manual in Frescobaldi
* single pages when clicking on a Help button in a dialog
* the manual on a website

What do you think?

While we're at it, what would you think of generally displaying the userguide in a Tool panel (as part of the "Viewers" subset) rather than a separate MainWindow? This would still require a pop-up solution when clicking on Help in a dialog, but it might be more consistent to have it as a tool, say, next to the documentation browser.
It could also be subject to a Preference, so `F1` would either open/switch to a Tool or open the separate userguide window.

**Comment by wbsoft** [comment]
I designed the help window to be floating and not too large indeed so that it works als from dialogs etc. 

I also wanted to code a html export facility to make it easy to add the user guide to a website.

But using an established system for organising and rendering the built-in userguide seems a good idea to me!

**Comment by uliska** [comment]
I must say that the current (visual) design will make it difficult to handle when the content grows significantly. Both regarding the navigation and pages with much content.

By now I'd say:

- have the HTML content generated by a site generator 
- embed it as a Tool panel
- have dialoh Help buttons open the requested page in a floating window (should basically be the same widget that can be reused)

**Comment by uliska** [comment]
I haven't really read into this, but wouldn't it also make sense to consider the [Qt Help Framework](https://doc.qt.io/qt-5/qthelp-framework.html) as a "native" approach to the task? It should equally be possible to export a help system as a separate website from there, if I'm not mistaken

**Comment by wbsoft** [comment]
Yes that's a very good idea I think. We should make sure that it's easy to extract translatable strings for the i18n.

**Comment by wbsoft** [comment]
Concerning the dock widget idea: I still believe that a separate top-level help window, not interfering with the application, is the most user friendly solution.

**Comment by uliska** [comment]
> I still believe that a separate top-level help window, not interfering with the application, is the most user friendly solution.

Well, maybe my issue is actually more about the actual *display* in that window, which doesn't seem very scalable (in terms of adding and managing more content). So changing that to a dock widget is certainly the part of the issue with the lowest priority for me.

(And BTW it should be nearly trivial to provide both options with a preference. Setting up a dock widget is little effort, and AFAICS one would simply load the identical widget in either context)

**Comment by uliska** [comment]
> We should make sure that it's easy to extract translatable strings for the i18n.

Yes. But without further reading I would *assume* that this is a topic where Qt provides excellent support.

**Comment by uliska** [comment]
Obviously I closed this by accident.

**Comment by uliska** [comment]
Ok, reading a *little bit* of the introduction reveals:

*  Qt Help is *only* about packaging documentation and
  * either providing it for reading in Qt Assistant
  * or displaying it in an application
* It does *not* deal with the generation and appearance of the documentation

So we would still need to set up a static site generator to *first* create a full HTML site,*then* process it with the Qt Help generator, and finally adapt the help handling to open help.

I suggest the way to go forward would be:

1) choose a SSG
2) set up the templates for a nice user guide
3) write a script for one-time use to copy as much as possible of the existing content to the new system
4) write a script or repeatedly process the documentation (including uploading it somewhere for an online manual)
5) rewrite the way Frescobaldi accesses the manual and replace the existing `userguide` module.

@fedelibre would you be able to help us out with 1) and 2) ?

**Comment by fedelibre** [comment]
Yes, I may help with 1) and 2). I'm very busy these days but things should be slightly better next week...

I would use the following approach to be sure to get to a working site ASAP.
I'd use an *existing theme* and customize it just a bit the appearance (e.g. color scheme). Then I would focus on enabling all the features we need for the site (deployment, translation,...) so we can start adding content.
Eventually when the website is drafted we may consider creating a new theme (the most time consuming and difficult part), unless we're happy with what we have. All the SSG have a clear distinction between content and theming so changing theme along the way is not a problem.

How to choose the SSG which would fit our needs?
I know and like mostly Mkdocs and Hugo. Probably the best way to take a decision together is creating a dummy site with each of these tools and see pros and cons for us.
I'll keep you posted.

**Comment by uliska** [comment]
Yes, that approach sounds reasonable. Adding content should ideally be done by a script, if that's possible. Have the existing files parsed by the existing `userguide` module (because that already knows how to read them) and then output them in an alternative format (which hopefully works out automatically).

I have two comments/ideas on the *design* part of things:

* We should *not* let ourselves be influenced by the *current* page design in the user guide because that is pretty deficient at the moment. Rather we should look at existing websites with documentation and think about that. readthedocs.org ([python-ly](https://python-ly.readthedocs.io/en/latest/index.html)) for example gives very clear documentation. Qt (https://doc.qt.io/qt-5/) is somewhat less intuitive, and especially way too complex for Frescobaldi's needs. I don't like [Pandoc's](https://pandoc.org/MANUAL.html) manual too much either, but these are just a few.
* We should think about a separation of concerns, although it isn't currently present in the userguide: There should be some sort of introductory material and then a reference. The current contents are *mostly* of that introductory nature, with *a few* pages trying to document functionality in more detail. This also makes it somewhat difficult to add documentation for new features, because a) it's often unclear *where* it should go and b) the pages don't provide enough space for detailed explanations.

With regard to the SSG I would (given your suggestions) vote for Mkdocs because it is Python-based. That might prove useful when we might need to add our own additions. On first glance I see that Mkdocs has a plugin API - we could for example use this to integrate python-ly's syntax highlighting.
And I think it might be useful to (at least optionally) allow Frescobaldi itself to update the manual, which should be easier when it is a Python tool. The repository is a 10 MB download so I wouldn't include it in Frescobaldi and also not add it as a requirement. But we could find a way for people running Frescobaldi from Git to make use of it, either by detecting it upon start or by providing an entry in Preferences=>Helper Applications. This would even allow Frescobaldi to update the documentation while running, although that's not *that* necessary.

**Comment by fedelibre** [comment]
> Adding content should ideally be done by a script, if that's possible. Have the existing files parsed by the existing `userguide` module (because that already knows how to read them) and then output them in an alternative format (which hopefully works out automatically).

I'm not sure I got it.
Do you mean that Markdown files for the documentation should be imported from the userguide module of Frescobaldi repository into the new doc repository?
I thought it should be the other way around, for a number of reasons.

**Comment by fedelibre** [comment]
I've just put up a *stub* of the website in [this repository](https://github.com/fedelibre/frescobaldi.org-mkdocs).I did not put any effort in it, that's clear :-)
I just wanted to give a concrete idea of what Mkdocs can or can't do.

Please read the notes I wrote in the README.
The main limitation of Mkdocs is that translating the content is not supported. Mkdocs team is quite small and seems not interested in this feature...
Mkdocs does not support content variables (in markdown) either. Probably not essential, but good to have. Maybe or maybe not: a simple Markdown file may be better, as it can be read and understood by any software.

On the other hand, Hugo supports content translation, content variables (called shortcodes) and has a big and active team. But no way to preprocess markdown files. If we need to preprocess markdown files only to use `ly` for syntax highlighting, we may find other solutions for this matter.

I may sketch another stub using Hugo, if it helps.

**Comment by uliska** [comment]
Thank you for preparing this.
I couldn't go into depth for all the resources but I had a look at the repository, the compiled website and (a cursory look) at the issue discussions over at the MkDocs repository.

The site itself already looks good, even when so closely based on the preset theme (which of course it the purpose of such themes). I think right now it somewhat blurs the line between a *website* and a *manual*, but that's something we could easily sort out.

I *think* that missing support for variables is not *that* much of a problem. These are currently used as a helper for some sort of standardization, and probably their use cases should be possible to handle with the tools of any SSG (and presumably there should be a way to convert the existing files appropriately).

The idea of creating a Pygments definition based on `python-ly`'s knowledge is something we should pursue. See also https://github.com/frescobaldi/python-ly/issues/116 where it had already been brought up.
I don't know if there would be some merit in considering to actually *use* Pygments as a highlighting engine for Frescobaldi.

What *is* a big issue AFAICS is the lack of translation support. We can't have a Frescobaldi manual without proper internationalization.

The issue with Hugo would be adding an “alien” set of dependencies. But to what extent is this actually an issue?

* Probably using Hugo requires installing Go and Hugo on anyone's computer who wants to *build* the manual (correct?)
* But a Python based system *also* requires installation of extra software (just not Python).
* As we probably don't want to add the *rendered* HTML files to the Git repository this would mean anyone using Frescobaldi from Git would need to have Hugo installed in order to use the manual. This looks somewhat problematic - but is not related to Hugo at all, it is an implication of the idea of using an SSG in the first place. A workaround for this would be to point the user to an online version of the manual - which is something numerous programs do by default nowadays (although I don't really like that).
* The most serious issue I can see with Hugo is the fact that it uses a different language, and that if anything has to be done in the code involving Go we'd need to have someone available who is able to deal with it. While I expect anyone working on Frescobaldi to be *able* to learn that it is something that adds to the threshold and (especially considering how small a contributor community we are) raises the risk of stuff just not being done.

**Comment by fedelibre** [comment]
>     * Probably using Hugo requires installing Go and Hugo on anyone's computer who wants to _build_ the manual (correct?)

No. The nice thing of Go applications is that they can be distributed as a single binary.
So all you need to get Hugo is downloading a binary file (or install one package from the repository if you use a Linux distro).

I see "using Frescobaldi from Git" as a developer way of using the application to test a new patch. In the last years it became a normal way for many users (including me) of using a recent version, but this happened only because we've not had a release for such a long time.
I write this only to say that IMO there's nothing wrong in having a "limited" version when running from source. I don't mean to say that it _should_ be limited. All the building may be automated by a build tool, which compiles the manual, updates the translations, etc. So a full-featured experience would be possible using one or two commands before launching the application.

Consider that Hugo doesn't have a translation based on gettext (po files).
The different sites are managed [using markdown files](https://gohugo.io/content-management/multilingual/).
AFAIK there's no official tool to check the differences between a translation and the original, but I guess this can be easily automated using Git (as in lilypond documentation).

Need more time to think about it...

**Comment by uliska** [comment]
> but this happened only because we've not had a release for such a long time.

That may be true, but I personally would like to see more and more people using Frescobaldi from Git because we need these people to *test* new features before merging. And in the end it's not much effort to do so, isn't it?

> Consider that Hugo doesn't have a translation based on gettext (po files).
The different sites are managed using markdown files.

I don't know what the effort would be to migrate the existing translations. But probably it would be possible to basically *render* the userguide pages like they are rendered within Frescobaldi and then harvest the results (which would be a one-time process).
Currently the translations are stored in message strings in the po files, which of course gives a more fine-grained relation between paragraphs and their translations. OTOH I would even find it more convenient to maintain them in Markdown files. I find the translation interface somewhat deterring, so that might even lead to more people giving a helping hand.

**Comment by uliska** [comment]
Just throwing in some other idea/question: Is it possible with reasonable effort to have Pandoc produce HTML documentation from a hierarchy of Markdown files?

It is reasonably simple to inject arbitrary functionality with "filters" (processing functions that can be written in Haskell, Lua, or Python (Haskell/Lua are executed natively while Python would be called externally)).

I'm just thinking of that because it would provide an easy path to professional PDF manuals too.

**Comment by fedelibre** [comment]
It would be more flexible but it needs much more time and study to set it up.

I don't know what the effort would be and I don't have time to dig it through right now. I see there are some [user contributed HTML templates](https://github.com/jgm/pandoc/wiki/User-contributed-templates#html).

**Comment by fedelibre** [comment]
There's a pretty nice example of website built using Pandoc [here](https://github.com/wcaleb/website).

**Comment by fedelibre** [comment]
> > Adding content should ideally be done by a script, if that's possible. Have the existing files parsed by the existing `userguide` module (because that already knows how to read them) and then output them in an alternative format (which hopefully works out automatically).
> 
> I'm not sure I got it.
> Do you mean that Markdown files for the documentation should be imported from the userguide module of Frescobaldi repository into the new doc repository?
> I thought it should be the other way around, for a number of reasons.

I take back my comment above.

As long as the markdown files from the userguide module are simple markdown (they are not right now), we may import them within the website repository using git submodule (no matter which tool we decide to use for the website).

This would mean that:

- developers adding new features or updating old ones will be able to document it working on a single repository
- user guide will still be available when running Fresco from source
- the website repository would be meant only to work on website pages (not for the manual)

I hope to get back to this issue in a couple of weeks, when I've finished some pending stuff for Frescobaldi.

**Comment by uliska** [comment]
Some comments on that:

* The original idea behind this issue was exactly to *replace* the current `userguide` module with something that provides advanced formatting, i.e. the SSG.
* My plan was to - once the page templates have been done - to convert the current `userguide` pseudo-Markdown files to whatever the SSG expects (the more native the Markdown the better).
* I would accept the backside that the manual has to be generated when anything changes, which affects people running Frescobaldi from the Git repository.
* The converted Markdown files will still be stored inside the Frescobaldi repository, either in the same directory or in a newly created one.

**Comment by fedelibre** [comment]
> The idea of creating a Pygments definition based on `python-ly`'s knowledge is something we should pursue. See also [frescobaldi/python-ly#116](https://github.com/frescobaldi/python-ly/issues/116) where it had already been brought up.

For the records, Pandoc does *not* use Pygments but a Haskell library which *should* have support for lilypond syntax:
- https://pandoc.org/MANUAL.html#syntax-highlighting
- https://github.com/jgm/skylighting
- https://github.com/KDE/syntax-highlighting/blob/master/data/syntax/lilypond.xml

I cannot test it now.

PS
@uliska I think you accidentally set a wrong milestone here.

**Comment by uliska** [comment]
I must have mixed up something. Pygments is used by the minted LaTeX package, and for some reason I thought this was in the Pandoc part of my toolchain.

Pandoc *does* support the LilyPond language, but not as well as python-ly.

**Comment by uliska** [comment]
OK, I have only skimmed through that long discussion, but I have had some exposure to MkDocs in the meantime, including writing and contributing to plugins.

I would now vote for generating the user guide as a website with MkDocs, integrate it into the Frescobaldi website and the user guide within Frescobaldi. Which interface (floating/tool/dialog) is a separate issue whose discussion should be not mixed up here.

### Localisation

This is the most important and (at least to me) most unclear topic.

My suggestion is to have a preprocessor that extracts all the localizations and writes markdown files to generate one website for each language. Only *then* MkDocs will be invoked. 
It would be possible to use [mkdocs-library](https://github.com/uliska/mkdocs-library) to render the translations as "partial books" like I will do with package documentation on https://openlilylib.org.

Right now I'm not fully clear how the userguide is actually translated. I see the call to `_(text)` at the end of `userguide/read.py`, but I don't see how translators know what to translate and where to put it.

Maybe it would be easier to manage translations on file level. While that may be less exact it might OTOH be much easier to contribute to. That would involve several steps/layers:

We'd have one directory per language, with English being the reference.

If for a given page there is a corresponding page in a translation directory we'll use that. Let's check whether it is up to date (last commit newer than the last commit touching the English file), and if not add a visual indicator on the page (with a link to the English version).

Have a tool (either a script or something within Frescobaldi when it is run from Git) that checks the files for a language and lists missing or out-of-date files, so contributors know what to update. 

The downside to that approach is of course that translators have to do the comparison manually. As a remedy I suggest that the mentioned tool provides access to the Git diff of the English file since the last version.

### LilyPond Highlighting

I have in the meantime started writing a MkDocs plugin that uses `python-ly` for LilyPond highlighting. It won't be an issue making that robust enough to be used as a (minor) PyPi dependency.

### Variables etc.

I haven't looked into it right now, but I'm sure we'll be able to do anything we might need with MkDocs plugins (which we can easily write).

### Integration in the regular website

I think the user guide should be added as a part in the regular Frescobaldi website. I have again looked at Federico's sketch at https://fedelibre.github.io/frescobaldi.org-mkdocs/ and would propose going in that direction, but using the Material theme's `tabs` option to achieve something like this:

![grafik](https://user-images.githubusercontent.com/1812148/76405411-5cb09580-6388-11ea-8e52-57fd00434999.png)

**Comment by fedelibre** [comment]
> Right now I'm not fully clear how the userguide is actually translated. I see the call to `_(text)` at the end of `userguide/read.py`, but I don't see how translators know what to translate and where to put it.

See [here](https://github.com/frescobaldi/frescobaldi/blob/master/i18n/update-pot.py#L62) and the i18n/ directory.
Translators just update a PO file.

**Comment by fedelibre** [comment]
> Currently the translations are stored in message strings in the po files, which of course gives a more fine-grained relation between paragraphs and their translations. OTOH I would even find it more convenient to maintain them in Markdown files. I find the translation interface somewhat deterring, so that might even lead to more people giving a helping hand.

I have translation experience with both PO files and the "multiple files" way you mention above (which is used by LilyPond documentation and website). My opinion is that PO files are way easier for a number of reasons:

- gettext is a "standard" and translators are used to it
- relation between paragraph and their translations is very important to keep the translation consistent (you can easily check how you translated a word before)
- updating a content by looking at a diff, then searching the line matching that diff in the localized file, etc. is very time consuming

There's one advantage in using a diff-like method: you can see what's changed in the original string.
Actually, you can see it also in a PO file if you use a regular text editor (as a fuzzy string keeps the old english string commented, until it's marked as not fuzzy). But things are changing: POedit already shows a diff of a fuzzy string, while Gtranslator is probably going to implement it soon (development is active again...).

Anyway, I would suggest to postpone the problem of translating the website once the english website is ready. It's better having an updated english-only website than an old localized website.

**Comment by wbsoft** [comment]
A side note: I think it is a perfectly valid idea to build the userguide with sphinx, instead of a local markdown-like system like it is now.

There is one little drawback: currently some contents of the user guide is retrieved from the running application, like keyboard shortcuts. Those adjust in the docs when the user has adjusted the shortcut in the preferences. But that's no real show stopper.

So I propose to convert at some time in the future the userguide to Sphinx. While I created [parce](https://parce.info) I became proficient in it. Maintaining documentation is very easy, as is building a pot file for the documentation. And exporting to a website (with nice templates) or PDF.

**Comment by wbsoft** [comment]
But the MKdocs website version also looks very slick! What do we do? :-)

**Comment by fedelibre** [comment]
If Mkdocs supported i18n, I would choose it without any doubt.
But as it doesn't, I'm open to consider Sphinx.

> Adding to the previous: also for the whole Frescobaldi website, Sphinx makes perfectly sense.

It would be better to have a single tool for both website and doc.

Default Sphinx themes are not responsive, as  far as I can see.
But there are nice third party themes for Sphinx:

- https://sphinx-themes.org/
- https://github.com/bashtage/sphinx-material/ (this is the same theme used in above Mkdocs sketch)

**Comment by Jean-Abou-Samra** [comment]
Now that Pygments has a LilyPond lexer, there is almost no special setup to get highlighted LilyPond source code in Sphinx documentation. All you need is a recent version of Pygments, which is as easy as

```
https://github.com/pygments/pygments/archive/master.zip
```

in the `requirements.txt` file.

**Comment by fedelibre** [comment]
@wbsoft If there's consensus on using Sphinx, why don't you add a **frescobaldi/frescobaldi.org** repository with a basic Sphinx (version 4) sketch? Then we might discuss there how to write the site and allow anyone to contribute.

If we break the work in many small tasks, I might consider contributing when I have some spare time.

**Comment by Jean-Abou-Samra** [comment]
I'm not sure why @uliska and @fedelibre originally preferred MkDocs, but it might be because of Markdown vs. reStructuredText syntax. In that case, be aware that these days it is possible to use Sphinx with Markdown.

https://myst-parser.readthedocs.io/en/latest/

More precisely, it's MyST (MarkedlyStructuredText) which is a kind of Markdown with reST-inspired extensibility.

That might also lower the work needed to convert the existing Markdown documentation.

**Comment by fedelibre** [comment]
Great work! I'll have more time to test it next week.

Just a quick question: how translations will be managed? Your screenshot seems to suggest that this has not been implemented yet.

**Comment by fedelibre** [comment]
If I understand correctly, the QtWebEngine problem on macOS might be solved if we change the tool to create the .dmg installer? (as discussed on #1143 )

**Comment by fedelibre** [comment]
I think QtWebEngine works fine on macOS in current 4.x packages.

We are about to get access to frescobaldi.org finally, so it's a good moment to revive this effort.
